build-fhs-userenv-bubblewrap: Preserve symlinks in /etc

zhaofengli opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

If the original file in `/etc` is a symlink, make it a symlink inside the sandbox as well. This PR is a more complete solution to https://github.com/NixOS/nixpkgs/issues/126234#issuecomment-1192203498, replacing #145258.

If you bind-mount an individual file and the file is deleted then recreated, the mountpoint will still refer to the original file and cannot be further bind-mounted:

```bash
touch mp1 mp2
echo "original content" >file
sudo mount -o bind file mp1
cat mp1 # "original content"

# delete and recreate the file
rm file
echo "new content" >file
cat mp1 # "original content"

# the steam soldier runtime uses bwrap internally, causing another bind-mount
sudo mount -o bind mp1 mp2 # "mount(2) system call failed: No such file or directory."
```

Since both `/etc/static` and `/run` are bind-mounted already, we should just keep the symlinks inside the sandbox.

Tested in the same manner as in https://github.com/NixOS/nixpkgs/issues/126234#issuecomment-1192203498. I can launch games after disconnecting and reconnecting to Wi-Fi with `steam -steamdeck`.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

queezle42 reviewed (approved):
This looks like a solid fix for https://github.com/NixOS/nixpkgs/issues/126234 and similar issues.

LGTM

SuperSandro2000 reviewed (commented):
We can now drop https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/editors/vscode/generic.nix#L177-L179 right?

sg-qwt commented:
I believe so. It'd also solve many other trick issues caused by missing symlink in sandbox (eg, wrong timezone in steam client)

lheckemann commented on pkgs/build-support/build-fhs-userenv-bubblewrap/default.nix:
```diff
@@ -126,9 +127,21 @@ let
       done
     fi
 
+    for i in ${etcBindEntries}; do
```
```suggestion
    for i in ${lib.escapeShellArgs etcBindEntries}; do
```

lheckemann commented on pkgs/build-support/build-fhs-userenv-bubblewrap/default.nix:
```diff
@@ -71,8 +71,8 @@ let
     ];
-  in concatStringsSep "\n  "
-  (map (file: "--ro-bind-try $(${coreutils}/bin/readlink -m /etc/${file}) /etc/${file}") files);
+  in concatStringsSep " "
+    (map (file: "\"/etc/${file}\"") files);
```
If you apply the escapeShellArgs suggestion...
```suggestion
  in map (path: "/etc/${path}") files;
```
(changed `file` to `path` since some of them aren't regular files at all)

lheckemann reviewed (approved):
I like this, some non-blocking nits though.

zhaofengli committed:
build-fhs-userenv-bubblewrap: Preserve symlinks in /etc

If the original file in /etc is a symlink, make it a symlink inside
the sandbox as well.

This fixes https://github.com/NixOS/nixpkgs/issues/126234#issuecomment-1192203498

Co-authored-by: Linus Heckemann <git@sphalerite.org>

zhaofengli committed:
build-fhs-userenv-bubblewrap: Use more descriptive names

hellwolf commented:
Back link a related fix in https://github.com/NixOS/nixpkgs/pull/286360
